### DISCUSSION TITLE
Ambiguous Error: json.hpp version 3.11.3

### DISCUSSION THREAD

**Discussion by nivihelen99** [nlohmann/json#4431] [Q&A]
json.hpp:7108:58: error: ISO C++ says that these are ambiguous, even though the worst conversion for the first is better than the worst conversion for the second: [-Werror]
 7108 |             for (auto it = ref_stack.back()->begin(); it != ref_stack.back()->end(); ++it)
      |                                                       ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~

I get the above error when I include json.hpp in to my project. Any idea of how to solve this? I am fairly new to C++

**Comment by nivihelen99** [comment]
More error logs:
json.hpp: In instantiation of 'bool nlohmann::json_abi_v3_11_3::detail::json_sax_dom_callback_parser<BasicJsonType>::end_object() [with BasicJsonType = nlohmann::json_abi_v3_11_3::basic_json<>]':
/usr/include/swss/json.hpp:12373:33:   required from 'bool nlohmann::json_abi_v3_11_3::detail::parser<BasicJsonType, InputAdapterType>::sax_parse_internal(SAX*) [with SAX = nlohmann::json_abi_v3_11_3::detail::json_sax_dom_callback_parser<nlohmann::json_abi_v3_11_3::basic_json<> >; BasicJsonType = nlohmann::json_abi_v3_11_3::basic_json<>; InputAdapterType = nlohmann::json_abi_v3_11_3::detail::iterator_input_adapter<const char*>]'
/usr/include/swss/json.hpp:12266:31:   required from 'void nlohmann::json_abi_v3_11_3::detail::parser<BasicJsonType, InputAdapterType>::parse(bool, BasicJsonType&) [with BasicJsonType = nlohmann::json_abi_v3_11_3::basic_json<>; InputAdapterType = nlohmann::json_abi_v3_11_3::detail::iterator_input_adapter<const char*>]'
/usr/include/swss/json.hpp:23338:118:   required from 'static nlohmann::json_abi_v3_11_3::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType, CustomBaseClass> nlohmann::json_abi_v3_11_3::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType, CustomBaseClass>::parse(IteratorType, IteratorType, nlohmann::json_abi_v3_11_3::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType, CustomBaseClass>::parser_callback_t, bool, bool) [with IteratorType = const char*; ObjectType = std::map; ArrayType = std::vector; StringType = std::__cxx11::basic_string<char>; BooleanType = bool; NumberIntegerType = long int; NumberUnsignedType = long unsigned int; NumberFloatType = double; AllocatorType = std::allocator; JSONSerializer = nlohmann::json_abi_v3_11_3::adl_serializer; BinaryType = std::vector<unsigned char>; CustomBaseClass = void; nlohmann::json_abi_v3_11_3::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType, CustomBaseClass>::parser_callback_t = std::function<bool(int, nlohmann::json_abi_v3_11_3::detail::parse_event_t, nlohmann::json_abi_v3_11_3::basic_json<>&)>]'
/usr/include/swss/json.hpp:24475:42:   required from here
/usr/include/swss/json.hpp:7108:58: error: ISO C++ says that these are ambiguous, even though the worst conversion for the first is better than the worst conversion for the second: [-Werror]
 7108 |             for (auto it = ref_stack.back()->begin(); it != ref_stack.back()->end(); ++it)
      |                                                       ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/10/utility:69,
                 from /usr/include/c++/10/tuple:38,
                 from /usr/include/c++/10/bits/hashtable_policy.h:34,
                 from /usr/include/c++/10/bits/hashtable.h:35,
                 from /usr/include/c++/10/unordered_map:46,
                 from mirrororch.cpp:3:
/usr/include/c++/10/bits/stl_relops.h:87:7: note: candidate 1: 'bool std::rel_ops::operator!=(const _Tp&, const _Tp&) [with _Tp = nlohmann::json_abi_v3_11_3::detail::iter_impl<nlohmann::json_abi_v3_11_3::basic_json<> >]'
   87 |       operator!=(const _Tp& __x, const _Tp& __y)
      |       ^~~~~~~~
In file included from portsorch.h:19,
                 from mirrororch.h:6,
                 from mirrororch.cpp:12:
/usr/include/swss/json.hpp:13373:10: note: candidate 2: 'bool nlohmann::json_abi_v3_11_3::detail::iter_impl<BasicJsonType>::operator!=(const IterImpl&) const [with IterImpl = nlohmann::json_abi_v3_11_3::detail::iter_impl<nlohmann::json_abi_v3_11_3::basic_json<> >; typename std::enable_if<(std::is_same<IterImpl, nlohmann::json_abi_v3_11_3::detail::iter_impl<BasicJsonType> >::value || std::is_same<IterImpl, nlohmann::json_abi_v3_11_3::detail::iter_impl<typename std::conditional<std::is_const<_Tp>::value, typename std::remove_const<_Tp>::type, const BasicJsonType>::type> >::value), std::nullptr_t>::type <anonymous> = nullptr; BasicJsonType = nlohmann::json_abi_v3_11_3::basic_json<>]'
13373 |     bool operator!=(const IterImpl& other) const

**Comment by nivihelen99** [comment]
using namespace std::rel_ops;
It appears the if I have this in my CPP file the above issue happens

**Comment by nlohmann** [comment]
Which compiler and OS are you using?

**Reply by wozza-lgx** [reply]
Hi, I'm getting this too:

gcc (GCC) 11.5.0 20240719 (Red Hat 11.5.0-2.0.1)
Oracle Enterprise Linux 9.5
Targeting: -std=c++20

Thanks

**Reply by nlohmann** [reply]
Strange. We use GCC 11.5.0 on Ubuntu 22.04.1 LTS in our CI. Can you compile and run the test suite?
